Use `sphinx.ext.linkcode` for more precise source code links

melechlapson opened the pull request:
## Summary
This addresses https://github.com/Qiskit/documentation/issues/517 by switching out the sphinx.ext.viewcode for the sphinx.ext.linkcode extension which allows our GitHub links in the documentation to be linked to the specific lines of code, not just the file. This was tested successfully in https://github.com/Qiskit/qiskit_sphinx_theme/pull/589 so now we want to implement it in the qiskit, runtime, and provider repos.

Example links generated in this PR build:

- a class definition https://github.com/Qiskit/qiskit/tree/main/qiskit/circuit/controlflow/break_loop.py#L21-L47
- a method https://github.com/Qiskit/qiskit/tree/main/qiskit/circuit/instruction.py#L229-L269
- a function https://github.com/Qiskit/qiskit/tree/main/qiskit/utils/deprecation.py#L104-L204

The API generation script at qiskit/documentation already knows how to handle `sphinx.ext.linkcode`.
 
## Determining the GitHub branch

We need to detect when to use `main` vs stable branches like `stable/1.0`, particularly for stable docs builds with GitHub Actions that get used by qiskit/documentation to deploy the docs. PR and local builds are less critical because the docs don't get published and are only for previews by code authors.

We do this with GitHub Actions environment variables. The docs workflow sets the env variable `QISKIT_DOCS_GITHUB_BRANCH_NAME` read by Sphinx in `conf.py`. You can see the workflow correctly passing the value to Sphinx [in this example run](https://github.com/Eric-Arellano/qiskit-terra/actions/runs/8194996708/job/22412261427#step:6:202). 

Examples of the workflow logic:

```
❯ GITHUB_REF_NAME="1.0.1rc1" bash test.sh  # tag
Using branch 'stable/1.0' for GitHub source code links
❯ GITHUB_REF_NAME="stable/0.2" bash test.sh     # branch build   
Using branch 'stable/0.2' for GitHub source code links
❯ GITHUB_BASE_REF=main bash test.sh  # PR
Using branch 'main' for GitHub source code links
```

### PR builds use `main`

PR builds and the merge queue use Azure Pipelines. To keep this infrastructure simpler, we simply default to `main` in PR builds, rather than adding Azure support. 

This means the URL will go to the wrong branch in docs previews for PRs against stable branches, although the stable branch build will work correctly. I don't expect this to matter much because PR builds are solely for docs previews; I'm skeptical code authors will be opening up the `[source]` link to double check the line numbers are exactly correct, since that line number calculation is 100% automated.

melechlapson committed:
switched from sphinx.ext.viewcode to sphinx.ext.linkcode

melechlapson committed:
removed extra line

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/Qiskit/qiskit?pullRequest=11851) <br/>All committers have signed the CLA.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/Qiskit/qiskit?pullRequest=11851) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/Qiskit/qiskit?pullRequest=11851) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/Qiskit/qiskit?pullRequest=11851) it.</sub>

melechlapson committed:
Merge branch 'main' into mlapson/switch-to-sphinx-ext-linkcode

Eric-Arellano commented on docs/conf.py:
```diff
@@ -15,9 +15,11 @@
 # pylint: disable=invalid-name,missing-function-docstring
 
 """Sphinx documentation builder."""
-
```
Nit: It's best practice to avoid irrelevant changes in PRs because those changes add noise. You can see your changes with `git diff`. Before opening a PR, I also like to review GitHub's diff before I hit submit.

melechlapson replied:
probably mistakenly hit backspace. I'll re-add it

Eric-Arellano commented on docs/conf.py:
```diff
@@ -155,3 +157,41 @@
 plot_html_show_formats = False
+
+
+# Need this method for the sphinx.ext.linkcode extension
```
```suggestion
# ----------------------------------------------------------------------------------
# Source code links
# ----------------------------------------------------------------------------------
```

Eric-Arellano commented on docs/conf.py:
```diff
@@ -155,3 +157,41 @@
+def linkcode_resolve(domain, info):
+    """
+    Determine the URL corresponding to Python object
+    """
```
Imo this comment doesn't add much. The code is already self-documenting.
```suggestion
```

melechlapson replied:
leftover from when I copied the method from the numpy example

Eric-Arellano commented on docs/conf.py:
```diff
@@ -155,3 +157,41 @@
+        ending_lineno = lineno + len(source) - 1
+        linespec = f"#L{lineno}-L{ending_lineno}"
+
+    return f"https://github.com/Qiskit/qiskit_sphinx_theme/tree/{release}/{file_name}/{linespec}"
```
1. Needs to be Qiskit
2. The `/` between the file and linespec is not necessary
3. Using the release is wrong; in that case, it would be `stable/{version}`. But we still need to figure out how we're going to detect `main` vs branch. In the meantime, let's set the `branch_name` variable. We'll fill in the other edge cases soon.

```suggestion
    git_branch = "main"

    return f"https://github.com/Qiskit/qiskit/tree/{git_branch}/{file_name}{linespec}"
```

melechlapson committed:
Add section header for source code links

Co-authored-by: Eric Arellano <14852634+Eric-Arellano@users.noreply.github.com>

melechlapson committed:
removed docstring

Co-authored-by: Eric Arellano <14852634+Eric-Arellano@users.noreply.github.com>

melechlapson committed:
update return string

Co-authored-by: Eric Arellano <14852634+Eric-Arellano@users.noreply.github.com>

melechlapson committed:
added back blank line

melechlapson committed:
Merge branch 'mlapson/switch-to-sphinx-ext-linkcode' of https://github.com/melechlapson/qiskit into mlapson/switch-to-sphinx-ext-linkcode

melechlapson committed:
Added a method to determine the GitHub branch

Co-authored-by: Eric Arellano <14852634+Eric-Arellano@users.noreply.github.com>

Eric-Arellano commented on docs/conf.py:
```diff
@@ -155,3 +160,67 @@
+# ----------------------------------------------------------------------------------
+# Source code links
+# ----------------------------------------------------------------------------------
+def determine_github_branch() -> str:
```
```suggestion

def determine_github_branch() -> str:
```

Eric-Arellano commented on docs/conf.py:
```diff
@@ -155,3 +160,67 @@
+    """Determine the GitHub branch name to use for source code links. We need to decide whether to use `stable/<version>` vs. `main` for dev builds.
+     Refer to https://docs.github.com/en/actions/learn-github-actions/variables for how we determine this with GitHub Actions.
+    """
+    print(os.environ)
```
```suggestion
```

melechlapson committed:
Merge branch 'main' into mlapson/switch-to-sphinx-ext-linkcode

melechlapson committed:
add blank line

Co-authored-by: Eric Arellano <14852634+Eric-Arellano@users.noreply.github.com>

melechlapson committed:
remove print statement

Co-authored-by: Eric Arellano <14852634+Eric-Arellano@users.noreply.github.com>

Eric-Arellano committed:
Try to fix error for contextlib file

We got this stacktrace:

Traceback (most recent call last):
  File "/home/vsts/work/1/s/.tox/docs/lib/python3.8/site-packages/sphinx/events.py", line 96, in emit
    results.append(listener.handler(self.app, *args))
  File "/home/vsts/work/1/s/.tox/docs/lib/python3.8/site-packages/sphinx/ext/linkcode.py", line 55, in doctree_read
    uri = resolve_target(domain, info)
  File "/home/vsts/work/1/s/docs/conf.py", line 216, in linkcode_resolve
    file_name = PurePath(full_file_name).relative_to(repo_root)
  File "/opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/pathlib.py", line 908, in relative_to
    raise ValueError("{!r} does not start with {!r}"
ValueError: '/opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/contextlib.py' does not start with '/home/vsts/work/1/s'

We shouldn't even attempt to generate a link for the file contextlib.py

melechlapson committed:
Try to fix error for Jenkins run #20240221.52

New build failed with this stacktrace:
Traceback (most recent call last):
  File "/home/vsts/work/1/s/.tox/docs/lib/python3.8/site-packages/sphinx/events.py", line 96, in emit
    results.append(listener.handler(self.app, *args))
  File "/home/vsts/work/1/s/.tox/docs/lib/python3.8/site-packages/sphinx/ext/linkcode.py", line 55, in doctree_read
    uri = resolve_target(domain, info)
  File "/home/vsts/work/1/s/docs/conf.py", line 215, in linkcode_resolve
    full_file_name = inspect.getsourcefile(obj)
  File "/opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/inspect.py", line 696, in getsourcefile
    filename = getfile(object)
  File "/opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/inspect.py", line 665, in getfile
    raise TypeError('{!r} is a built-in class'.format(object))
TypeError: <class 'builtins.CustomClassical'> is a built-in class

So I added a condition that the obj should not be a built-in

Eric-Arellano committed:
Check that qiskit in module name sooner

melechlapson committed:
moved valid code object verification earlier

melechlapson committed:
Merge branch 'main' into mlapson/switch-to-sphinx-ext-linkcode

melechlapson committed:
added try except statement to getattr call

melechlapson committed:
added extra try/except block

Eric-Arellano committed:
Also support Azure Pipelines

melechlapson committed:
removed unused import

Eric-Arellano committed:
Revert Azure support to keep things simple

melechlapson committed:
added extra "/" to final URL

melechlapson committed:
Merge branch 'mlapson/switch-to-sphinx-ext-linkcode' of https://github.com/melechlapson/qiskit into mlapson/switch-to-sphinx-ext-linkcode

melechlapson committed:
Merge branch 'main' into mlapson/switch-to-sphinx-ext-linkcode

Eric-Arellano committed:
Merge branch 'main' of github.com:Qiskit/qiskit-terra into mlapson/switch-to-sphinx-ext-linkcode

Eric-Arellano committed:
Move GitHub branch logic to GitHub Action

Eric-Arellano committed:
Merge branch 'mlapson/switch-to-sphinx-ext-linkcode' of github.com:melechlapson/qiskit into mlapson/switch-to-sphinx-ext-linkcode

mtreinish commented on docs/conf.py:
```diff
@@ -155,3 +159,48 @@
+            inspect.isclass(obj) or inspect.ismethod(obj) or inspect.isfunction(obj)
+        )
+        if not is_valid_code_object:
+            return None
```
How does this work for attributes? Does it not matter because sphinx doesn't generate source links for documented attributes of a class.

Eric-Arellano replied:
Yeah, attributes are skipped.

mtreinish commented on .github/workflows/docs_deploy.yml:
```diff
@@ -19,19 +19,30 @@ jobs:
+          elif [[ $GITHUB_REF_NAME =~ ^([0-9]+\.[0-9]+) ]]; then
+            BRANCH_NAME="stable/${BASH_REMATCH[1]}"
+          else
+              BRANCH_NAME="$GITHUB_REF_NAME"
```
I know that @jakelishman has strong opinions about using the github actions built-in variables here (see: https://github.com/Qiskit/qiskit/pull/11901 ). But personally I think this is ok; but the concerns about manual builds are valid and I wonder how this works if you want to do a local build for debugging that points to the correct place.

Eric-Arellano replied:
Jake didn't want the GitHub actions to live inside `conf.py`. He wasn't opposed to using GitHub Actions built-in variables in general, only where those are set. He suggested at the bottom of #11901 setting the values in the CI files like this.

For local builds, we default to `main`. You can override the value by locally setting `QISKIT_DOCS_GITHUB_BRANCH_NAME=stable/1.0`. I don't think it's worth the complexity to try to do something clever. I'd rather we keep it simple, at least until we hear devs banging on the doors that this is a problem for them and they need auto-detection for local builds.

jakelishman replied:
I'll just resolve this to move it out the way. For completeness' sake, Eric's right: I was complaining about using the GitHub environment variables within `docs/conf.py`, especially since the PR I was complaining about was trying to make it work with Azure as well - I just wanted to see the CI-specific environment variables being resolved into Qiskit-specific forms in the CI-specific configuration files, which link CI to Qiskit, which is now what's being done.

mtreinish commented on docs/conf.py:
```diff
@@ -155,3 +159,48 @@
+            inspect.isclass(obj) or inspect.ismethod(obj) or inspect.isfunction(obj)
+        )
+        if not is_valid_code_object:
+            return None
```
I'm thinking we should just drop this logic altogether. Could we just use `importlib` to import the object directly? At the end of the day this is just trying to get the fully qualified path without doing an import at all. I also worry about manually accessing `sys.modules` directly

Eric-Arellano replied:
I'll try switching to importlib. This current code is taken from how numpy does it and we didn't try importlib.

melechlapson replied:
@mtreinish is this better?

melechlapson committed:
switched to importlib and removed redundant block of code

Eric-Arellano commented on docs/conf.py:
```diff
@@ -155,3 +159,43 @@
+    module_name = info["module"]
+    module = importlib.import_module(module_name)
+    if module is None or "qiskit" not in module_name:
+        return None
```
```suggestion
    module_name = info["module"]
    if "qiskit" not in module_name:
        return None
     
    try: 
        module = importlib.import_module(module_name)
    except ModuleNotFoundError:
        return None
```

Eric-Arellano commented on docs/conf.py:
```diff
@@ -11,13 +11,17 @@
 # that they have been altered from the originals.
 
 from __future__ import annotations
```
```suggestion
from __future__ import annotations

```

melechlapson committed:
Apply suggestions from code review

Co-authored-by: Eric Arellano <14852634+Eric-Arellano@users.noreply.github.com>

melechlapson committed:
Merge branch 'main' into mlapson/switch-to-sphinx-ext-linkcode

jakelishman commented on .github/workflows/docs_deploy.yml:
```diff
@@ -19,19 +19,30 @@ jobs:
           # Sync with 'documentationPythonVersion' in 'azure-pipelines.yml'.
           python-version: '3.9'
-
       - name: Install dependencies
```
Minor nit: please can we restore the spacing lines between the job stages?  There's another deletion below the `tox -e docs` line as well.

jakelishman commented on .github/workflows/docs_deploy.yml:
```diff
@@ -19,19 +19,30 @@ jobs:
+          elif [[ $GITHUB_REF_NAME =~ ^([0-9]+\.[0-9]+) ]]; then
+            BRANCH_NAME="stable/${BASH_REMATCH[1]}"
+          else
+              BRANCH_NAME="$GITHUB_REF_NAME"
```
- Minor: is there a reason to use the old-style `[` operator in the first test but the new-style bash `[[` in the second?
- For PRs, do you want to set the base ref to the PR target, or would it be more appropriate to use the commit hash of the temporary PR merge commit?  I'm just thinking that PRs will end up generating incorrect ranges if they change any code, because they'll base the decision about lines of code based on the actual PR, but then output a link to the unchanged `main`.  If you use `GITHUB_SHA` you can probably get an accurate link. 
- What's the reason for having a tag point at the branch it's on and not at the tree of the tag itself?  Seems like the tag would be more reliably accurate.

Eric-Arellano replied:
I removed the PR conditional. I realized it's not even relevant to this workflow because it never runs on PRs, which we use Azure for. It was bad bash-ified copy pasta from Runtime. 

Re: PR builds in general, see the PR description for why I think it's fine to at least start with using `main`. If devs do end up actually clicking the link and being confused, then we could add the appropriate logic to Azure in a follow up.

--

Re: tags pointing at branches, this is helpful for us in qiskit/documentation to have a smaller diff when regenerating docs for the same minor release, such as between a tag release vs. `workflow_dispatch`. We try to check the diff when generating docs to make sure there is nothing surprising. Changing every GitHub link is noisy.

jakelishman commented on docs/conf.py:
```diff
@@ -155,3 +160,47 @@
+        return None
+    if full_file_name is None:
+        return None
+    file_name = full_file_name.split("/qiskit/")[-1]
```
It's quite unlikely to matter, but you might prefer to do this `full_file_name.split` thing using the proper `os.path` or `pathlib` tooling.  Something like:

```python
from pathlib import Path

REPO_ROOT = Path(__file__).resolve().parents[1]

def linkcode_resolve(domain, info):
    # ...

    file_name = str(Path(full_file_name).resolve().relative_to(REPO_ROOT / "qiskit"))
```

The current form would break if we were ever to add a subdirectory of the `qiskit` package that itself was called `qiskit` - unlikely, but not impossible.

jakelishman replied:
Oh, actually not even unlikely - I just remembered that I literally proposed doing that in #10737 (though in that case, there wouldn't be any Python-documentable objects in it).

Eric-Arellano replied:
Thanks for the suggestion!

Eric-Arellano replied:
Ah, I now remember why we were using `[-1]`. This current code results in the relative file path `.tox/docs/lib/python3.8/site-packages/qiskit/circuit/classicalfunction/exceptions.py`. 

I'm unclear why that is: I couldn't reproduce in `qiskit_sphinx_theme` with setting in `tox.ini` `isolated_build = true` and `usedevelop = false` to mirror Qiskit. (I've been having issues running tox locally in Qiskit due to autodoc). In `qiskit_sphinx_theme` locally and GitHub Actions, the path is the normal file path you'd expect without the `.tox` prefix. So I think we want this code to support both styles.

We could use a regex replace to remove the `.tox` prefix if it's there.

jakelishman replied:
In that case, perhaps you want to use `import qiskit; qiskit.__file__` as the file-path root, so it'll automatically find the base of where the files are installed?

I don't mind a huge amount, though.

jakelishman reviewed (commented):
Thanks for this, and sorry about the delay.  I just left a couple of minor nits/questions in line, but otherwise I'm happy to move to merge.

melechlapson committed:
added back spaces

Eric-Arellano committed:
Merge branch 'main' of github.com:Qiskit/qiskit-terra into mlapson/switch-to-sphinx-ext-linkcode

Eric-Arellano committed:
Clarify docs_deploy GitHub logic

1. Remove misleading PR conditional. This worfklow doesn't even run in PRs. It was bad copy-pasta from the runtime repo
2. Clarify why we point tag builds to their stable branch name

Eric-Arellano committed:
Use pathlib for relativizing file name

jakelishman reviewed (dismissed):
Thanks for the changes, this seems fine to me.

Eric-Arellano committed:
Fix relative_to() path

Eric-Arellano committed:
Remove tox prefix

Eric-Arellano commented:
This is now generating correct links in the PR preview, e.g. https://github.com/Qiskit/qiskit/blob/main/qiskit/circuit/gate.py#L169-L226. Should be ready to merge

jakelishman reviewed (approved):
imo https://github.com/Qiskit/qiskit/pull/11851#discussion_r1553600536 is less fragile than the regex path-match here, but also, as long as it's working for you right now, I can't see the regex replace actually _breaking_ the docs build, so I'm ok just to merge things
